File: maximum-building-height.cc
// Maximum Building Height
class Solution {
public:
  int maxBuilding(int n, vector<vector<int>>& res) {
    vector<pair<int, int>> a;
    a.emplace_back(1, 0);
    a.emplace_back(n, n-1);
    for (auto &x: res)
      a.emplace_back(x[0], x[1]);
    sort(a.begin(), a.end());
    int m = a.size(), ans = 0;
    for (int i = 0; i < m-1; i++) {
      int d = a[i+1].first-a[i].first;
      a[i+1].second = min(a[i+1].second, a[i].second+d);
    }
    for (int i = m; --i; ) {
      int d = a[i].first-a[i-1].first;
      a[i-1].second = min(a[i-1].second, a[i].second+d);
      ans = max(ans, a[i-1].second+a[i].second+d >> 1);
    }
    return ans;
  }
};

File: maximum-building-height.rs
impl Solution {
    pub fn max_building(n: i32, restrictions: Vec<Vec<i32>>) -> i32 {
        use std::cmp;
        let mut a: Vec<(i32, i32)> = restrictions.into_iter().map(|x| (x[0], x[1])).collect();
        a.push((1, 0));
        a.push((n, n-1));
        a.sort_unstable();
        let mut ans = 0;
        for i in 0..a.len()-1 {
            a[i+1].1 = a[i+1].1.min(a[i+1].0-a[i].0+a[i].1);
        }
        for i in (1..a.len()).rev() {
            let d = a[i].0-a[i-1].0;
            a[i-1].1 = a[i-1].1.min(d+a[i].1);
            ans = ans.max(a[i-1].1+a[i].1+d >> 1);
        }
        ans
    }
}
